ppl/zqd/recruiter.go

```diff
@@ -44,32 +37,68 @@ func handleRegister(c *Core, w http.ResponseWriter, r *http.Request) {
 	if !request(c, w, r, &req) {
 		return
 	}
-	registered, err := c.workerPool.Register(req.Addr, req.NodeName)
-	if err != nil {
-		respondError(c, w, r, zqe.ErrInvalid(err))
+	if req.Timeout <= 0 {
+		respondError(c, w, r, zqe.E(zqe.Invalid, "required parameter timeout"))
 		return
 	}
-	respond(c, w, r, http.StatusOK, api.RegisterResponse{
-		Registered: registered,
-	})
-}
-
-func handleUnreserve(c *Core, w http.ResponseWriter, r *http.Request) {
-	var req api.UnreserveRequest
-	if !request(c, w, r, &req) {
+	timer := time.NewTimer(time.Duration(req.Timeout) * time.Millisecond)
```

Comment:
The changes here are an improvement to address previous comments about recruitment & idle timeout. I think the code would be less complex if it was handled inside the Recruit call and what is now the 'Register' call. If you combine the Register/Deregister calls into a single call, like 'recruiter.WaitForRecruitment(ctx, req.Addr, req.NodeName, req.Timeout)'.

Reply:
I like that idea. I will take a look at how to do it.

Reply:
I removed a redundant call to workerPool.Deregister which made this code confusing. An important aspect of the design is that workerPool.Recruit automatically deregisters recruited workers. The explicit Deregister call is only needed in the case of a request timeout or context cancel. So, workerPool.Deregister does not always follow workerPool.Register, which was the way the previous code made it look.

The code in the workerPool methods is as minimal as possible to avoid holding the mutex lock longer than needed. I can introduce an intermediate layer between the request handlers in the zqd package and the workerPool methods if you think that will make the code more readable.

Reply:
yes please; I think an intermediate layer would be helpful to separate the api handling from the register goroutine wakeup details.

Reply:
I moved this code into recruiter/longpoll.go. It looks better to me; the code works the same, but less cluttered.
I like having the channel as a local variable, so it is easy to find the readers and writers.